keps/sig-windows/2258-node-service-log-viewer/README.md

```diff
@@ -0,0 +1,393 @@
+issues will require being on the node, issues with the kube-proxy or kubelet,
+to name a couple, could be solved by perusing their logs. However this
+too requires the administrator to SSH access into the nodes. Having a way for
+them to view the logs using kubectl will significantly simplify their
+troubleshooting.
+
+
+### Goals
+Provide a cluster administrator with a streaming view of logs using kubectl
+without them having to implement a client side reader or logging into the node.
+This would work for:
+- Services on Linux worker and control plane nodes:
+  - That have systemd / journald support.
+  - That have services that log to `/var/log/`
+- Windows worker nodes (all supported variants) that log to `C:\var\log`,
+  System and Application logs, Windows Event Logs and Event Tracing (ETW).
+
+### Non-Goals
+- Providing support for non-systemd Linux distributions.
+- Reporting logs for nodes that have config or connection issues with the
+  cluster.
+- Getting logs from services that do not use /var/log/.
+
+## Proposal
+
+### Implement client for logs endpoint viewer (OS agnostic)
+- Extend `kubectl logs` to work with node objects.
+- Implement a client for the `/var/log/` kubelet endpoint viewer.
+
+### Linux distros with systemd / journald
+Supplement the the `/var/log/` endpoint viewer on the kubelet with a thin shim
+over the `journal` directory that shells out to journalctl. Then extend
+`kubectl logs` to also work with node objects.
+
+### Linux distributions without systemd / journald
+Running the new "kubectl logs nodes" command against services on nodes that do
+not use systemd / journald should return "OS not supported". However getting
+logs from `/var/log/` should work on all systems.
+
+### Windows
+Reuse the kubelet API for querying the Linux journal for invoking the
+`Get-WinEvent` cmdlet in a PowerShell.
+
+### User Stories
+
+Consider a scenario where pods / containers are refusing to come up on certain
+nodes. As mentioned in the motivation section, troubleshooting this scenario
+involves the cluster administrator to SSH into nodes to scan the logs. Allowing
+them to use `kubectl logs` to do the same as they would to debug issues with a
+pod / container would greatly simply their debug workflow. This also opens up
+opportunities for tooling and simplifying automated log gathering. The feature
+can also be used to debug issues with Kubernetes services especially in Windows
+nodes that run as native Windows services and not as DaemonSets or Deployments.
+
+Here are some example of how a cluser administrator would use this feature:
+```
+# See what logs are available in masters in /var/log/
+kubectl logs nodes --role master --path=/
+
+# Display cron log file (/var/log/cron) from all masters
+kubectl logs nodes --role master --path=cron
+
+# Show kubelet and crio journal logs from all masters
+kubectl logs nodes --role master --path journal -s kubelet -s crio
```

Comment:
The "unify" flag which was opt-in did the zipper (linked in the other comment) which is what I wanted called out as a separate note.  The behavior with unify was "dump".  Unify got added because most people *loved* bulk node log fetching but then hated doing | sort -n 1 -k correctly and dealing with sort issues.  I will note there are gaps in unify - early in system boot journal doesn't have hostname so you get each node reporting "localhost" and from the output of sort you CANNOT tell which one is which, but with unify you know which node you're talking to and thus can annotate those.

Reply:
I'm more asking about the contract - is this simply "you get what you get" or is there an expectation.  I missed the `unify` flag, which implies that journal is zippered by default, which seems right.